src/datahike/writing.cljc

```diff
@@ -86,57 +109,63 @@
-  (let [{:keys [hash max-tx max-eid config]
-         {:keys [parents]} :meta} db]
-    (if (:crypto-hash? config)
-      (uuid [hash max-tx max-eid parents])
-      (uuid))))
-
-(defn commit! [store config db parents]
-  (let [parents (or parents #{(get config :branch)})
-        parents (branch-heads-as-commits store parents)
-        cid (create-commit-id db)
-        db (-> db
-               (assoc-in [:meta :datahike/commit-id] cid)
-               (assoc-in [:meta :datahike/parents] parents))
-        db-to-store (db->stored db true)]
-    (k/assoc store cid db-to-store {:sync? true})
-    (k/assoc store (:branch config) db-to-store {:sync? true})
-    db))
-
-(defn update-and-commit!
-  ([connection tx-data tx-meta update-fn]
-   (update-and-commit! connection tx-data tx-meta update-fn nil))
-  ([connection tx-data tx-meta update-fn parents]
-   (let [{:keys [db/noCommit]} tx-meta
-         {:keys [db-after]
-          {:keys [db/txInstant]}
-          :tx-meta
-          :as   tx-report}     (update-fn connection tx-data tx-meta)
-         {:keys [config]} db-after
-         {:keys [store writer]} @(:wrapped-atom connection)
-         new-meta               (assoc (:meta db-after) :datahike/updated-at txInstant)
-         db                     (assoc db-after :meta new-meta :writer writer)
-         db                     (if noCommit db (commit! store config db parents))
-         tx-report              (assoc tx-report :db-after db)
-         tx-report              (if noCommit
-                                  tx-report
-                                  (assoc-in tx-report [:tx-meta :db/commitId]
-                                            (get-in db [:meta :datahike/commit-id])))]
-     (reset! connection db)
-     (doseq [[_ callback] (some-> (:listeners (meta connection)) (deref))]
-       (callback tx-report))
-     tx-report)))
+  (let [{:keys [hash max-tx max-eid meta]} db]
+    (uuid [hash max-tx max-eid meta])))
+
+(defn commit!
+  ([db parents]
+   (commit! db parents true))
+  ([db parents sync?]
+   (async+sync sync? *default-sync-translation*
+               (go-try-
+                (let [{:keys [store config]} db
+                      parents       (or parents #{(get config :branch)})
+                      parents       (branch-heads-as-commits store parents)
+                      cid           (create-commit-id db)
+                      db            (-> db
+                                        (assoc-in [:meta :datahike/commit-id] cid)
+                                        (assoc-in [:meta :datahike/parents] parents))
+                      db-to-store   (db->stored db true)
+                      _             (<?- (flush-pending-writes store sync?))
+                      commit-log-op (k/assoc store cid db-to-store {:sync? sync?})
+                      branch-op     (k/assoc store (:branch config) db-to-store {:sync? sync?})]
+                  (<?- commit-log-op)
+                  (<?- branch-op)
```

Comment:
This is waiting for the write operations in the asynchronous case and a no-op in the synchronous case (because the values are already written then). It also will re-throw exceptions that would be in either of these channels.